flink-ml-lib/src/main/java/org/apache/flink/ml/feature/minmaxscaler/MinMaxScalerModel.java

```diff
@@ -0,0 +1,187 @@
+                                            broadcastModelKey,
+                                            getMax(),
+                                            getMin(),
+                                            getFeaturesCol()),
+                                    outputTypeInfo);
+                        });
+        return new Table[] {tEnv.fromDataStream(output)};
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+        ReadWriteUtils.saveModelData(
+                MinMaxScalerModelData.getModelDataStream(modelDataTable),
+                path,
+                new MinMaxScalerModelData.ModelDataEncoder());
+    }
+
+    /**
+     * Loads model data from path.
+     *
+     * @param env Stream execution environment.
+     * @param path Model path.
+     * @return MinMaxScalerModel model.
+     */
+    public static MinMaxScalerModel load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+        MinMaxScalerModel model = ReadWriteUtils.loadStageParam(path);
+        DataStream<MinMaxScalerModelData> modelData =
+                ReadWriteUtils.loadModelData(
+                        env, path, new MinMaxScalerModelData.ModelDataDecoder());
+        return model.setModelData(tEnv.fromDataStream(modelData));
+    }
+
+    /** This operator loads model data and predicts result. */
+    private static class PredictOutputFunction extends RichMapFunction<Row, Row> {
+        private final String featureCol;
+        private final String broadcastKey;
+        private final double upperBound;
+        private final double lowerBound;
+        private DenseVector scaleVector;
+        private DenseVector offsetVector;
+
+        public PredictOutputFunction(
+                String broadcastKey, double upperBound, double lowerBound, String featureCol) {
+            this.upperBound = upperBound;
+            this.lowerBound = lowerBound;
+            this.broadcastKey = broadcastKey;
+            this.featureCol = featureCol;
+        }
+
+        @Override
+        public Row map(Row row) {
+            if (scaleVector == null) {
+                MinMaxScalerModelData minMaxScalerModelData =
+                        (MinMaxScalerModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+                DenseVector minVector = minMaxScalerModelData.minVector;
+                DenseVector maxVector = minMaxScalerModelData.maxVector;
+                scaleVector = new DenseVector(minVector.size());
+                offsetVector = new DenseVector(minVector.size());
+                for (int i = 0; i < maxVector.size(); ++i) {
+                    if ((minVector.values[i] - maxVector.values[i]) == 0.0) {
```

Comment:
According to the discussion in [1][2], it is could lead to unexpected bug if we use `==` to check whether two double values are equal.

How about we follow the recommendation in [3] and use the following code to check equality? The idea here is to explicitly specify an epsilon similar to what we did in tests.

```
Math.abs(minVector.values[i] - maxVector.values[i]) < 1e-5
```

[1] https://stackoverflow.com/questions/25160375/comparing-double-values-for-equality-in-java
[2] https://publicobject.com/2009/11/floating-point-equality.html
[3] https://www.baeldung.com/java-comparing-doubles

Reply:
I think, [3] is much better. I will refine the code later.